In [1]:
import pandas as pd

In [2]:
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

In [3]:

df=pd.read_csv('Melbourne_housing.csv')

In [4]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0   3/9/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2   4/2/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3   4/2/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4   4/3/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [5]:
df.shape

(34857, 21)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Price          27247 non-null  float64
 5   Method         34857 non-null  object 
 6   SellerG        34857 non-null  object 
 7   Date           34857 non-null  object 
 8   Distance       34856 non-null  float64
 9   Postcode       34856 non-null  float64
 10  Bedroom2       26640 non-null  float64
 11  Bathroom       26631 non-null  float64
 12  Car            26129 non-null  float64
 13  Landsize       23047 non-null  float64
 14  BuildingArea   13742 non-null  float64
 15  YearBuilt      15551 non-null  float64
 16  CouncilArea    34854 non-null  object 
 17  Lattitude      26881 non-null  float64
 18  Longti

## Functions for checking missing values
- The isnull() function is used for checking the missing values.
- The result is usually boolean values of True and False.
- Missing values get mapped to True
- Non-missing values get mapped to False

In [7]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

Our dataset has 34857 entries. To be more efficient in analyzing the missing columns we can get a percentage of the missing values



---

##  1 Deleting Rows with missing Data

If the record contains a few missing values, we can drop all the rows having any missing values.

In [8]:
df1=df.dropna(axis=0)

In [9]:
df1.shape

(8887, 21)

You can see the result above has deleted a lot of rows and hence , this is not the best way to handle our missing values in this dataset

## 2 Deleting Columns with missing Data

In [ ]:
# dropping all columns with missing values by specifying axis=1(Not the best option)
df2=df.dropna(axis=1)

In [ ]:
df2.isnull().sum()

## dropping columns by specifying threshold

Calculating the percentage of values is usually 

Percentage= (Value/Total Value)×100

In [10]:
percentage_missing=df.isnull().sum()*100/len(df)

In [11]:
percentage_missing

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.832057
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.002869
Postcode          0.002869
Bedroom2         23.573457
Bathroom         23.599277
Car              25.039447
Landsize         33.881286
BuildingArea     60.576068
YearBuilt        55.386293
CouncilArea       0.008607
Lattitude        22.882061
Longtitude       22.882061
Regionname        0.008607
Propertycount     0.008607
dtype: float64

In [12]:
to_drop =percentage_missing[percentage_missing>55].keys()

In [13]:
to_drop

Index(['BuildingArea', 'YearBuilt'], dtype='object')

In [14]:
 df.drop(to_drop, axis=1,inplace=True)

In [15]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

BuildingArea and YearBuilt have been dropped

In [35]:
df.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            1
Postcode            1
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
CouncilArea         0
Lattitude        7976
Longtitude       7976
Regionname          0
Propertycount       0
dtype: int64

## 3 Replacing Missing Data with mean, mode or average using fillna function

In [23]:
## Replace missing values in Price with Mean 
df['Price'].fillna(df['Price'].mean(), inplace=True)

In [25]:
## Replace missing values in Bedroom2 with Mode
df['Bedroom2'].fillna(df['Bedroom2'].mode()[0],inplace=True)

In [27]:
## Replace missing values in Car with Median
df['Car'].fillna(df['Car'].median(),inplace=True)

In [28]:
# specifying a value directly
df['Propertycount'].fillna(value=0,inplace=True)

# 4 Replacing Missing values using Sklearn Impute(Categorical data)

In [31]:
from sklearn.impute import SimpleImputer

In [32]:

# replacing missing values for region with the most frequent value

imp_mode = SimpleImputer(strategy='most_frequent')


df['Regionname']= imp_mode.fit_transform(df[['Regionname']])

In [34]:
# replacing missing values in councilarea with value other


imp_mode = SimpleImputer(strategy='constant', fill_value='other')


df['CouncilArea']= imp_mode.fit_transform(df[['CouncilArea']])



In [36]:
# confirm 'other' has been filled for missing values
df['CouncilArea'].value_counts(ascending=True)

other                                3
Moorabool Shire Council              7
Mitchell Shire Council              20
Cardinia Shire Council              41
Macedon Ranges Shire Council        46
Nillumbik Shire Council             88
Yarra Ranges Shire Council         102
Casey City Council                 176
Frankston City Council             290
Melton City Council                292
Greater Dandenong City Council     314
Knox City Council                  371
Maroondah City Council             506
Whitehorse City Council            618
Wyndham City Council               624
Whittlesea City Council            828
Kingston City Council              871
Hobsons Bay City Council           942
Manningham City Council           1046
Yarra City Council                1186
Hume City Council                 1214
Port Phillip City Council         1280
Maribyrnong City Council          1451
Stonnington City Council          1460
Monash City Council               1466
Brimbank City Council    